haematite_models/src/irc/channel.rs

```diff
@@ -1,31 +1,113 @@
 use std::collections::{HashMap, HashSet};
 
-use linked_hash_set::LinkedHashSet;
-use serde::Serialize;
+use serde::{Serialize, Serializer};
 
+use super::error::Error;
+use super::membership::{Diff as MembershipDiff, Membership};
 use super::topic::Topic;
 
-#[derive(Default, Serialize)]
-pub struct Membership {
-    pub status: HashSet<char>,
-}
-
 #[derive(Default, Serialize)]
 pub struct Channel {
     pub topic: Option<Topic>,
     pub modes: HashMap<char, Option<String>>,
-    pub mode_lists: HashMap<char, LinkedHashSet<String>>,
+    pub mode_lists: HashMap<char, Vec<String>>,
     pub users: HashMap<String, Membership>,
 }
 
-impl Membership {
-    pub fn new() -> Self {
-        Self::default()
-    }
+pub enum Action<T> {
+    Add(T),
+    Remove,
+}
+
+pub enum Diff {
+    Topic(Option<Topic>),
+    Mode(char, Action<Option<String>>),
+    InternalModeList(char, Action<String>),
+    ExternalModeList(char, Vec<String>),
+    InternalUser(String, MembershipDiff),
+    ExternalUser(String, Action<Membership>),
 }
 
 impl Channel {
-    pub fn new() -> Self {
-        Self::default()
+    pub fn new(mode_lists: HashSet<char>) -> Self {
+        Self {
+            mode_lists: mode_lists.into_iter().map(|c| (c, Vec::new())).collect(),
+            ..Self::default()
+        }
+    }
+
+    /// # Errors
+    ///
+    /// Will return `Err` if the presented diff is not applicable to the
+    /// current network state, or if the result data cannot be serialized.
+    pub fn update<S>(&mut self, diff: Diff, ser: S) -> Result<(String, S::Ok), Error>
+    where
+        S: Serializer,
+    {
+        Ok(match diff {
+            Diff::Topic(topic) => {
+                self.topic = topic;
+                ("topic".to_owned(), self.topic.serialize(ser)?)
+            }
+            Diff::Mode(mode, action) => {
+                let value = match action {
+                    Action::Add(arg) => {
+                        let value = arg.serialize(ser)?;
+                        self.modes.insert(mode, arg);
+                        value
+                    }
+                    Action::Remove => {
+                        self.modes.remove(&mode);
+                        ser.serialize_none()?
+                    }
+                };
+                (format!("modes/{}", mode), value)
+            }
+            Diff::InternalModeList(mode, action) => {
+                let list = self.mode_lists.get_mut(&mode).ok_or(Error::UnknownMode)?;
+                let (index, value) = match action {
+                    Action::Add(arg) => {
+                        let value = arg.serialize(ser)?;
+                        list.push(arg);
+                        (list.len() - 1, value)
+                    }
+                    Action::Remove => {
+                        //let index = list.iter().position(|a| a == &arg).unwrap();
```

Comment:
this is commented out because `Action` doesn't have a way to hold a value, so I need to figure out different action enums